File: InterviewCampProblems/InterviewCampProblems.Code/Class14.cs
/*
Given a String, find the longest substring with unique characters. For example: "whatwhywhere" --> "atwhy"
*/
using System;
using System.Collections.Generic;

namespace InterviewCampProblems.Code {
    public class Class14 {

        public static int[] SlidingSubArrayString (string input) {

            if (string.IsNullOrEmpty (input)) {
                System.Console.WriteLine ("Null, Empty or Invalid string.");
                return null;
            }
            int[] outputarr = new int[2] { 0, 0 };
            char[] arr = input.ToCharArray ();
            int startIndex = 0;
            int endIndex = 0;
            int longest = 1;
            Dictionary<char, int> dict = new Dictionary<char, int> ();
            dict.Add (arr[0], 0);
            while (endIndex < arr.Length - 1) {
                endIndex++;
                if (dict.ContainsKey (arr[endIndex]) && dict[arr[endIndex]] >= startIndex) {
                    startIndex = dict[arr[endIndex]] + 1;
                    dict[arr[endIndex]] = endIndex;
                } else {
                    dict.Add (arr[endIndex], endIndex);
                }

                if (endIndex - startIndex + 1 > longest) {
                    longest = endIndex - startIndex + 1;
                    outputarr[0] = startIndex;
                    outputarr[1] = endIndex;
                }
            }

            return outputarr;

        }
    }
}

File: Java/InterviewCamp/src/main/java/Class14.java
//Given a String, find the longest substring with unique characters.For example: "whatwhywhere" --> "atwhy"

import com.sun.tools.javac.util.Pair;

import java.util.HashMap;

public class Class14 {

    public static int[] uniqueChars(String input)
    {
        if(input == null || input.length() == 0) {
            return null;
        }

        int first = 0, second = 0;
        int start = 0, end = 0, longest = 0;
        HashMap<Character,Integer> map = new HashMap<>();
        map.put(input.charAt(0),0);

        while (end < input.length()-1)
        {
            end++;
            char toAdd = input.charAt(end);
            if( map.containsKey(toAdd) && map.get(toAdd) >= start)
            {
                start = map.get(toAdd) + 1;
            }

            map.put(toAdd,end);

            if(end - start + 1 > longest)
            {
                longest = end - start + 1;
                first = start;
                second = end;
            }
        }

        return new int[] {first, second};

    }
}
